lib/SILOptimizer/Transforms/TempRValueElimination.cpp

```diff
@@ -601,6 +602,34 @@ void TempRValueOptPass::tryOptimizeCopyIntoTemp(CopyAddrInst *copyInst) {
 
   LLVM_DEBUG(llvm::dbgs() << "  Success: replace temp" << *tempObj);
 
+  // If copyInst's source must be deinitialized, whether that must be done via
+  // a newly created destroy_addr.
+  //
+  // If lastLoadInst is a load or a copy_addr, then the deinitialization can be
+  // done in that instruction.
+  //
+  // This is necessary for correctness: otherwise, copies of move-only values
+  // would be introduced.
+  bool needToInsertDestroy = [&]() {
+    if (!needFinalDeinit)
+      return false;
+    if (lastLoadInst == copyInst)
+      return true;
+    if (auto *cai = dyn_cast<CopyAddrInst>(lastLoadInst)) {
+      auto retval = cai->getSrc() != tempObj || !cai->isTakeOfSrc();
```

Comment:
nitpick: all these conditions are getting hard to understand. IMO, for clarity it's better to write
```
  if (cai->getSrc() == tempObj && cai->isTakeOfSrc()) {
    // Can reuse the last copy_addr as deinitialization.
    return false;
  } else {
    assert(!tempObj->getType().isMoveOnly() && "introducing copy of move-only value!?");
    return true;
  }
  ```

Reply:
https://github.com/apple/swift/pull/65099